quinn-proto/src/connection/mod.rs

```diff
@@ -2502,16 +2499,23 @@ impl Connection {
                         }
                     }
 
-                    if self.side.is_server() && self.peer_params.stateless_reset_token.is_none() {
-                        if self.rem_cids.active_seq() == 0 {
-                            // We're a server still using the initial remote CID for the client, so
-                            // let's switch immediately to enable clientside stateless resets.
-                            let _ = self.update_rem_cid();
-                        }
-                    } else if self.rem_cids.is_active_retired() {
-                        // If our current CID is meant to be retired; or
-                        // active remote CID is invalid (due to packet loss or reordering),
-                        // we must switch to next valid CID
+                    if !retired.is_empty() {
+                        self.spaces[SpaceId::Data]
+                            .pending
+                            .retire_cids
+                            .extend(retired);
+                        // The active CID was necessarily replaced, so we need to update our reset
+                        // token to match
+                        self.set_reset_token(
+                            self.rem_cids
+                                .active_reset_token()
+                                .expect("missing reset token for non-initial CID"),
+                        );
+                    }
+
+                    if self.side.is_server() && self.rem_cids.active_seq() == 0 {
+                        // We're a server still using the initial remote CID for the client, so
+                        // let's switch immediately to enable clientside stateless resets.
```

Comment:
Document the invariant for the `unwrap()` below a little more explicitly?

Reply:
Mooted by refactoring.